src/epoch.rs

```diff
@@ -0,0 +1,82 @@
+/// The global epoch
+///
+/// The last bit in this number is unused and is always zero. Every so often the global epoch is
+/// incremented, i.e. we say it "advances". A pinned mutator may advance the global epoch only if
+/// all currently pinned mutators have been pinned in the current epoch.
+///
+/// If an object became garbage in some epoch, then we can be sure that after two advancements no
+/// mutator will hold a reference to it. That is the crux of safe memory reclamation.
+
+use std::ops::Deref;
+use std::sync::atomic::AtomicUsize;
+use std::sync::atomic::Ordering::{Relaxed, Acquire, Release, SeqCst};
+
+use mutator::Registry;
+use mutator::Scope;
+use sync::list::{List, IterResult};
+use util::cache_padded::CachePadded;
+
+/// The global epoch is a (cache-padded) integer.
+#[derive(Default, Debug)]
+pub struct Epoch {
+    epoch: CachePadded<AtomicUsize>,
+}
+
+impl Epoch {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Attempts to advance the global epoch.
+    ///
+    /// The global epoch can advance only if all currently pinned mutators have been pinned in the
+    /// current epoch.
+    ///
+    /// Returns the current global epoch.
+    #[cold]
+    pub fn try_advance<'scope>(&self, registries: &List<Registry>, scope: &Scope) -> usize {
+        let epoch = self.epoch.load(Relaxed);
+        ::std::sync::atomic::fence(SeqCst);
+
+        // Traverse the linked list of mutator registries.
+        let mut registries = registries.iter(scope);
+        loop {
+            match registries.next() {
+                IterResult::Abort => {
+                    // We leave the job to the mutator that also tries to advance to epoch and
+                    // continues to iterate the registries.
+                    return epoch;
+                }
+                IterResult::None => break,
+                IterResult::Some(registry) => {
+                    let (mutator_is_pinned, mutator_epoch) = registry.get_state();
+
+                    // If the mutator was pinned in a different epoch, we cannot advance the global
+                    // epoch just yet.
+                    if mutator_is_pinned && mutator_epoch != epoch {
+                        return epoch;
+                    }
+                }
+            }
+        }
+        ::std::sync::atomic::fence(Acquire);
```

Comment:
Oh, I forgot to answer this question.  I am sorry.

This is in order to synchronize the release-writes to local epochs (https://github.com/crossbeam-rs/crossbeam-epoch/pull/4/files/42c2fb50e75ceedf71022173b539aa6f75e519d9#diff-1e3fb927a1f124f8fdede26bfa2a730dR211) via the relaxed-read of it (https://github.com/crossbeam-rs/crossbeam-epoch/pull/4/files/42c2fb50e75ceedf71022173b539aa6f75e519d9#diff-6d7fb2f2e4b00caaa3efe3c37b5708e9R52).